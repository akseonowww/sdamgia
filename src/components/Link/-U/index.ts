export { default } from './Link-U'
