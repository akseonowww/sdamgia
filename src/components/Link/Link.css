.Link,
.Link-U {
	font-weight: inherit;
	font-style: inherit;
	color: #006;

	border-bottom: 0.05rem solid;
	border-color: rgba(0, 0, 102, 0.25);
	text-decoration: none;

	transition-property: border-color, color, fill;
	transition-duration: 0.25s;
	transition-timing-function: ease-out;
}

.Link_black,
.Link_black-U {
	color: #000;

	border-bottom-color: rgba(0, 0, 0, 0.25);
}

.Link_selected {
	cursor: default;
}

.Link:not(.Link_static):not(.Link_black):not(.Link_pseudo):visited,
.Link-U:not(.Link_static-U):not(.Link_black-U):not(.Link_pseudo-U):visited {
	color: #8f00af;

	border-color: rgba(143, 0, 175, 0.25);
}

.Link_pseudo {
	margin: 0;
	padding: 0;

	font-family: inherit;
	font-size: inherit;
	text-align: left;

	background: transparent;
	border: none;

	outline: none;
}

.Link_pseudo,
.Link_pseudo-U {
	border-bottom-width: 0.1rem;
	border-bottom-style: dashed;
	border-color: rgba(9, 9, 73, 0.25);

	cursor: pointer;
}

.Link_pseudo-black,
.Link_pseudo-black-U {
	color: #000;

	border-color: rgba(0, 0, 0, 0.25);
}

.Link:hover {
	color: #f60 !important;

	border-color: rgba(255, 102, 0, 0.25) !important;

	transition: none;
}

.Link_pseudo:hover {
	border-color: rgba(255, 102, 0, 0.5) !important;
}

.Link_wrap {
	width: 100%;

	text-align: left;

	background: transparent;
	border: none;
}

.Link_wrap-U {
	color: inherit;

	text-decoration: none;

	transition: none;
}

.Link_wrap:hover .Link-U {
	border-color: rgba(255, 102, 0, 0.25) !important;

	transition: none;
}

.Link_wrap:hover .Link_pseudo-U {
	border-color: rgba(255, 102, 0, 0.5) !important;
}
