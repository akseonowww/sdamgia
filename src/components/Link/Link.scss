@import 'src/styles/variables', 'src/styles/transitions';

@mixin Link-Style {
	font-style: inherit;
	font-weight: inherit;
	color: $primary;
	text-decoration: none;

	border-bottom-width: $border-width;
	border-bottom-style: solid;
	border-color: $border-primary;

	@include transition;
}

@mixin Link_pseudo-Style {
	border-bottom-width: 0.1rem;
	border-bottom-style: dashed;
	border-color: rgba(9, 9, 73, 0.25);
	cursor: pointer;
}

.Link {
	@include Link-Style;

	&-U {
		@include Link-Style;
	}

	&_black {
		color: $black;

		border-bottom-color: $border-black;

		&-U {
			color: $black;

			border-bottom-color: $border-black;
		}
	}

	&_white {
		color: $white;

		&-U {
			border-bottom-color: $border-white;
		}
	}

	&:not(&_static):not(&_black):not(&_white):not(&_pseudo):visited,
	&-U:not(&_static-U):not(&_black-U):not(&_white-U):not(&_pseudo-U):visited {
		color: $visited;

		border-color: $border-visited;
	}

	&_pseudo {
		margin: 0;
		padding: 0;

		font-family: inherit;
		font-size: inherit;
		text-align: left;

		background: transparent;
		border: none;
		outline: none;

		@include Link_pseudo-Style;

		&-U {
			@include Link_pseudo-Style;
		}
	}

	&_pseudo-black {
		color: $black;

		border-color: $border-black;

		&-U {
			color: $black;

			border-color: $border-black;
		}
	}

	&_pseudo-white {
		color: $white;

		border-color: $border-white;

		&-U {
			color: $white;

			border-color: $border-white;
		}
	}

	&:hover {
		color: $danger !important;

		border-color: $border-danger !important;

		@include transition-reset;
	}

	&_white:hover {
		color: $white !important;

		border-bottom-color: $white !important;
	}

	&_pseudo:hover {
		border-color: $border-dangerous !important;
	}

	&_wrap {
		width: 100%;

		text-align: left;

		background: transparent;
		border: none;

		&-U {
			color: inherit;

			text-decoration: none;

			@include transition-reset;
		}

		&:hover {
			.Link-U {
				border-color: $border-danger !important;

				@include transition-reset;
			}

			.Link_white-U {
				border-color: $white !important;
			}

			.Link_pseudo-U {
				border-color: $border-dangerous !important;
			}
		}
	}
}
