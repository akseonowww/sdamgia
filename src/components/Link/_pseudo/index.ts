export { default } from './Link_pseudo'
