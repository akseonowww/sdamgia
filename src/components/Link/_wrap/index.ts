export { default } from './Link_wrap'
