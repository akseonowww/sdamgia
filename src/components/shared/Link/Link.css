.Link {
	color: #006;

	border-bottom: .75px solid;
	border-color: rgba(0, 0, 102, 0.25);
	text-decoration: none;
}

.Link_black {
	color: #000;

	border-bottom-color: rgba(0, 0, 0, 0.25);
}

.Link:not(.Link_static):not(.Link_black):not(.Link_pseudo):visited {
	color: #8f00af;

	border-color: rgba(143, 0, 175, 0.25);
}

.Link_pseudo {
	border-bottom-style: dashed;
	border-color: rgba(9, 9, 73, 0.5);

	cursor: pointer;
}

.Link_pseudo-black {
	color: #000;

	border-color: rgba(0, 0, 0, 0.5);
}

.Link:hover {
	color: #f60 !important;

	border-color: rgba(255, 102, 0, 0.25) !important;
}

.Link_pseudo:hover {
	border-color: rgba(255, 102, 0, 0.5) !important;
}
