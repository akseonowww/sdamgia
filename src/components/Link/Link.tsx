import React, { FC, ReactNode } from 'react'
import { Link as RouterLink, NavLink as RouterNavLink } from 'react-router-dom'
import cx from 'classnames'

import './Link.scss'
import './_selected/Link_selected.scss'

interface ILinkProps {
  className?: string
  classNameU?: string
  activeClassName?: string | boolean
  to?: string
  href?: string
  exact?: boolean
  title?: string
  handleClick?: () => void
  children: ReactNode
}

const Link: FC<ILinkProps> = ({
  className,
  classNameU,
  activeClassName,
  to,
  href = '/',
  title,
  exact = false,
  handleClick,
  children,
}) => {
  if (to && (activeClassName || exact)) {
    return (
      <RouterNavLink
        className={cx('Link', className)}
        activeClassName={cx('Link_selected', activeClassName)}
        to={to}
        exact={exact}
        title={title}
      >
        {children}
      </RouterNavLink>
    )
  }

  if (to) {
    return (
      <RouterLink className={cx('Link', className)} to={to} title={title}>
        {children}
      </RouterLink>
    )
  }

  if (handleClick && classNameU) {
    return (
      <div className={className} onClick={handleClick}>
        <u className={cx('Link-U', classNameU)}>{children}</u>
      </div>
    )
  }

  if (handleClick) {
    return (
      <div className={className} onClick={handleClick}>
        {children}
      </div>
    )
  }

  if (classNameU) {
    return (
      <a
        className={className}
        href={href}
        rel="noopener noreferrer"
        title={title}
      >
        <u className={cx('Link-U', classNameU)}>{children}</u>
      </a>
    )
  }

  return (
    <a
      className={cx('Link', className)}
      href={href}
      rel="noopener noreferrer"
      title={title}
    >
      {children}
    </a>
  )
}

export default Link
