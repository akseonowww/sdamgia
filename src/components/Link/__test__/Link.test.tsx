import React from 'react'
import { unmountComponentAtNode, render } from 'react-dom'
import { act } from 'react-dom/test-utils'
import pretty from 'pretty'
import { MemoryRouter } from 'react-router-dom'

import Link from '../Link'

describe('Link', () => {
  let container: HTMLElement | null = null
  const handleClick = jest.fn(() =>
    console.log('User clicked and saw less/more information')
  )

  beforeEach(() => {
    container = document.createElement('div')
    document.body.appendChild(container)
  })

  afterEach(() => {
    if (container) {
      unmountComponentAtNode(container)
      container.remove()
      container = null
    }
  })

  describe('renders 7 types:', () => {
    it('1. React Router Nav wrap', () => {
      act(() => {
        render(
          <MemoryRouter>
            <Link
              className="Link Link_wrap Link_black NavSide-Link"
              activeClassName
              to="/"
              exact
              title="Шкалы, статистика, оценивание заданий"
            >
              <u className="Link-U Link_wrap-U Link_black-U">About exam</u>
            </Link>
          </MemoryRouter>,
          container
        )
      })
      expect(container && container.textContent).toBe('About exam')
    })

    it('2. React Router Nav', () => {
      act(() => {
        render(
          <MemoryRouter>
            <Link className="Link_black Header-Link" activeClassName to="/">
              RESHU EGE
            </Link>
          </MemoryRouter>,
          container
        )
      })
      expect(container && container.textContent).toBe('RESHU EGE')

      act(() => {
        render(
          <MemoryRouter>
            <Link className="Link_black Header-Link" to="/" exact>
              RESHU EGE
            </Link>
          </MemoryRouter>,
          container
        )
      })
      expect(container && container.textContent).toBe('RESHU EGE')

      act(() => {
        render(
          <MemoryRouter>
            <Link
              className="Link_black Header-Link"
              activeClassName
              to="/"
              exact
            >
              RESHU EGE
            </Link>
          </MemoryRouter>,
          container
        )
      })
      expect(container && container.textContent).toBe('RESHU EGE')
    })

    it('3. React Router', () => {
      act(() => {
        render(
          <MemoryRouter>
            <Link className="Link_static" to="/archive">
              Archive
            </Link>
          </MemoryRouter>,
          container
        )
      })
      expect(container && container.textContent).toBe('Archive')
    })

    it('4. wrap with click handler', () => {
      act(() => {
        render(
          <Link
            className="Link_wrap Link_pseudo VariantBox-Control LarinVariants-Control"
            classNameU="Link_wrap-U Link_pseudo-U"
            handleClick={handleClick}
          >
            Show all
          </Link>,
          container
        )
      })
      expect(container && container.textContent).toBe('Show all')
    })

    it('5. with click handler', () => {
      act(() => {
        render(
          <Link className="Link_pseudo" handleClick={handleClick}>
            Solution
          </Link>,
          container
        )
      })
      expect(container && container.textContent).toBe('Solution')
    })

    it('6. wrap', () => {
      act(() => {
        render(
          <Link className="Link_wrap ProfileAuth-Link" href="/pass_change">
            Password Recovery
          </Link>,
          container
        )
      })
      expect(container && container.textContent).toBe('Password Recovery')
    })

    it('7. HTML', () => {
      act(() => {
        render(<Link>About Project</Link>, container)
      })
      expect(container && container.textContent).toBe('About Project')

      act(() => {
        render(<Link className="Link_black">About Project</Link>, container)
      })
      expect(container && container.textContent).toBe('About Project')
    })
  })

  describe('should render content of 7 types:', () => {
    it('1. React Router Nav wrap', () => {
      act(() => {
        render(
          <MemoryRouter>
            <Link
              className="Link Link_wrap Link_black NavSide-Link"
              activeClassName
              to="/"
              exact
              title="Шкалы, статистика, оценивание заданий"
            >
              <u className="Link-U Link_wrap-U Link_black-U">About exam</u>
            </Link>
          </MemoryRouter>,
          container
        )
      })
      if (container) expect(pretty(container.innerHTML)).toMatchSnapshot()
    })

    it('2. React Router Nav', () => {
      act(() => {
        render(
          <MemoryRouter>
            <Link className="Link_black Header-Link" activeClassName to="/">
              RESHU EGE
            </Link>
          </MemoryRouter>,
          container
        )
      })
      if (container) expect(pretty(container.innerHTML)).toMatchSnapshot()

      act(() => {
        render(
          <MemoryRouter>
            <Link
              className="Link_black Header-Link"
              activeClassName
              to="/"
              exact
            >
              RESHU EGE
            </Link>
          </MemoryRouter>,
          container
        )
      })
      if (container) expect(pretty(container.innerHTML)).toMatchSnapshot()

      act(() => {
        render(
          <MemoryRouter>
            <Link className="Link_black Header-Link" to="/" exact>
              RESHU EGE
            </Link>
          </MemoryRouter>,
          container
        )
      })
      if (container) expect(pretty(container.innerHTML)).toMatchSnapshot()
    })

    it('3. React Router', () => {
      act(() => {
        render(
          <MemoryRouter>
            <Link className="Link_static" to="/archive">
              Archive
            </Link>
          </MemoryRouter>,
          container
        )
      })
      if (container) expect(pretty(container.innerHTML)).toMatchSnapshot()
    })

    it('4. wrap with click handler', () => {
      act(() => {
        render(
          <Link
            className="Link_wrap Link_pseudo VariantBox-Control LarinVariants-Control"
            classNameU="Link_wrap-U Link_pseudo-U"
            handleClick={handleClick}
          >
            Show all
          </Link>,
          container
        )
      })
      if (container) expect(pretty(container.innerHTML)).toMatchSnapshot()
    })

    it('5. with click handler', () => {
      act(() => {
        render(
          <Link className="Link_pseudo" handleClick={handleClick}>
            Solution
          </Link>,
          container
        )
      })
      if (container) expect(pretty(container.innerHTML)).toMatchSnapshot()
    })

    it('6. wrap', () => {
      act(() => {
        render(
          <Link className="Link_wrap ProfileAuth-Link" href="/pass_change">
            Password Recovery
          </Link>,
          container
        )
      })
      if (container) expect(pretty(container.innerHTML)).toMatchSnapshot()
    })

    it('7. HTML', () => {
      act(() => {
        render(<Link>About Project</Link>, container)
      })
      if (container) expect(pretty(container.innerHTML)).toMatchSnapshot()

      act(() => {
        render(<Link className="Link_black">About Project</Link>, container)
      })
      if (container) expect(pretty(container.innerHTML)).toMatchSnapshot()
    })
  })
})
