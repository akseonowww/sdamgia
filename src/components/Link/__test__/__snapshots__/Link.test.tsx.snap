// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render content of 7 types: 1. React Router Nav wrap 1`] = `"<a aria-current=\\"page\\" class=\\"Link Link Link_wrap Link_black NavSide-Link Link_selected\\" title=\\"Шкалы, статистика, оценивание заданий\\" href=\\"/\\"><u class=\\"Link-U Link_wrap-U Link_black-U\\">About exam</u></a>"`;

exports[`Link should render content of 7 types: 2. React Router Nav 1`] = `"<a aria-current=\\"page\\" class=\\"Link Link_black Header-Link Link_selected\\" href=\\"/\\">RESHU EGE</a>"`;

exports[`Link should render content of 7 types: 2. React Router Nav 2`] = `"<a aria-current=\\"page\\" class=\\"Link Link_black Header-Link Link_selected\\" href=\\"/\\">RESHU EGE</a>"`;

exports[`Link should render content of 7 types: 2. React Router Nav 3`] = `"<a aria-current=\\"page\\" class=\\"Link Link_black Header-Link Link_selected\\" href=\\"/\\">RESHU EGE</a>"`;

exports[`Link should render content of 7 types: 3. React Router 1`] = `"<a class=\\"Link Link_static\\" href=\\"/archive\\">Archive</a>"`;

exports[`Link should render content of 7 types: 4. wrap with click handler 1`] = `"<div class=\\"Link_wrap Link_pseudo VariantBox-Control LarinVariants-Control\\"><u class=\\"Link-U Link_wrap-U Link_pseudo-U\\">Show all</u></div>"`;

exports[`Link should render content of 7 types: 5. with click handler 1`] = `"<div class=\\"Link_pseudo\\">Solution</div>"`;

exports[`Link should render content of 7 types: 6. wrap 1`] = `"<a class=\\"Link Link_wrap ProfileAuth-Link\\" href=\\"/pass_change\\" rel=\\"noopener noreferrer\\">Password Recovery</a>"`;

exports[`Link should render content of 7 types: 7. HTML 1`] = `"<a class=\\"Link\\" href=\\"/\\" rel=\\"noopener noreferrer\\">About Project</a>"`;

exports[`Link should render content of 7 types: 7. HTML 2`] = `"<a class=\\"Link Link_black\\" href=\\"/\\" rel=\\"noopener noreferrer\\">About Project</a>"`;
